hls4ml/templates/vivado/nnet_utils/nnet_graph.h

```diff
@@ -0,0 +1,439 @@
+        }
+      }
+    }
+
+    //output array --> output vec
+    nnet::mat_to_vec<res_T, res_T, typename CONFIG_T::edge_attr_aggr_config>(edge_attr_aggr, edge_attr_aggr_1D);
+  }
+
+  template<class data_T, class index_T, class res_T, typename CONFIG_T>
+    void edgeblock(
+            data_T    node_attr_1D[CONFIG_T::n_node*CONFIG_T::node_dim],
+			data_T    edge_attr_1D[CONFIG_T::n_edge*CONFIG_T::edge_dim],
+			index_T   edge_index_1D[CONFIG_T::n_edge*2],
+			res_T     edge_update_1D[CONFIG_T::n_edge*CONFIG_T::out_dim],
+			typename CONFIG_T::dense_config1::weight_t  core_edge_w0[CONFIG_T::dense_config1::n_in*CONFIG_T::dense_config1::n_out],
+			typename CONFIG_T::dense_config1::bias_t    core_edge_b0[CONFIG_T::dense_config1::n_out],
+			typename CONFIG_T::dense_config2::weight_t  core_edge_w1[CONFIG_T::dense_config2::n_in*CONFIG_T::dense_config2::n_out],
+			typename CONFIG_T::dense_config2::bias_t    core_edge_b1[CONFIG_T::dense_config2::n_out],
+			typename CONFIG_T::dense_config3::weight_t  core_edge_w2[CONFIG_T::dense_config3::n_in*CONFIG_T::dense_config3::n_out],
+			typename CONFIG_T::dense_config3::bias_t    core_edge_b2[CONFIG_T::dense_config3::n_out],
+			typename CONFIG_T::dense_config4::weight_t  core_edge_w3[CONFIG_T::dense_config4::n_in*CONFIG_T::dense_config4::n_out],
+			typename CONFIG_T::dense_config4::bias_t    core_edge_b3[CONFIG_T::dense_config4::n_out])
+  {
+    //initialize arrays
+    // 1. node_attr (input)
+    data_T node_attr[CONFIG_T::n_node][CONFIG_T::node_dim];
+    #pragma HLS ARRAY_PARTITION variable=node_attr complete dim=0
+    nnet::vec_to_mat<data_T, data_T, typename CONFIG_T::node_attr_config>(node_attr_1D, node_attr);
+
+    // 2. edge_attr (input)
+    data_T edge_attr[CONFIG_T::n_edge][CONFIG_T::edge_dim];
+    #pragma HLS ARRAY_PARTITION variable=edge_attr complete dim=0
+    nnet::vec_to_mat<data_T, data_T, typename CONFIG_T::edge_attr_config>(edge_attr_1D, edge_attr);
+
+    // 3. edge_index (input)
+    index_T edge_index[CONFIG_T::n_edge][2];
+    #pragma HLS ARRAY_PARTITION variable=edge_index complete dim=0
+    nnet::vec_to_mat<index_T, index_T, typename CONFIG_T::edge_index_config>(edge_index_1D, edge_index);
+
+    // 4. edge_update (output)
+    res_T edge_update[CONFIG_T::n_edge][CONFIG_T::out_dim];
+    #pragma HLS ARRAY_PARTITION variable=edge_update complete dim=0
+
+    int sender_col;
+    int receiver_col;
+    if(CONFIG_T::flow == source_to_target){
+      sender_col = 0;
+      receiver_col = 1;
+    }
+    else{
+      sender_col = 1;
+      receiver_col = 0;
+    }
+
+    #pragma HLS PIPELINE II=CONFIG_T::reuse_factor
+    edge_loop: for(int i = 0; i < CONFIG_T::n_edge; i++) { //for each edge
+      #pragma HLS UNROLL
+
+      // get sender, receiver indices
+      index_T s = edge_index[i][sender_col];
+      index_T r = edge_index[i][receiver_col];
+
+      // construct NN input: <receiver, sender, edge>
+      data_T node_concat[2*CONFIG_T::node_dim];

```

Comment:
You can save some registers and a couple of clocks here by concatenating three arrays into `phi_input` directly (by forgetting about `nnet::concatenate1d` and writing your own loop for concatenation)

Reply:
So I tested both a 1D and 2D implementation of this for the following configuration (this is kind of our "baseline" configuration):
- max-nodes=28
- max-edges=51
- n-neurons=8
- precision=ap_fixed<16,8>
- reuse=1

The synthesis yielded:
- Latency: +57% for both
- LUT usage: +87% for the 1d version, +58% for the 2d version
- FF usage: +169% for both
- DSP usage: -10% for both

So I think we should leave this as it is.

Reply:
Yep, it saves a few LUTs! Pushing the changes now